### PR TITLE
Implementation of web layouts

### PR DESCRIPTION
**Pull request by thekid**
This PR implements web layouts specified in xp-framework/rfc#261 

## New functionality
The configuration argument now handles directory references as well as layout and scriptlet class names, and a dash.

```sh
$ xpws -c de.thekid.dialog.WebLayout
# Will start the XP web server reading the web layout from the given layout class

$ xpws -c de.thekid.dialog.scriptlet.RssScriptlet
# Will start the XP web server with a web layout with the given scriptlet at "/"

$ xpws -c -
# Will start the XP web server to serve static files from document root
```

## Existing functionality
The following still works as expected

```sh
$ xpws -c etc
# Will start the XP web server and configure the web from etc/web.ini

$ xpws
# Will start the XP web server and configure the web from etc/web.ini if that exists,
# alternatively starting it to serve static files from document root
```

*Note: For Windows, this depends on xp-framework/xp-runners@12421ac1669356afa5bc898d009d446448e8a922 (released in 5.2.0)*

### CODE REVIEW

File: src/main/php/xp/scriptlet/Runner.class.php [hunk 1]
```diff
@@ -196,9 +204,17 @@ public function run($url= '/') {
         $instance->setTrace($cat);
       }
       $instance->init();
-    
+
+      // Set up request and response
+      $request= $instance->request();
+      $request->method= $_SERVER['REQUEST_METHOD'];
+      $request->env= $_ENV;
```
Comment (kiesel) [inline src/main/php/xp/scriptlet/Runner.class.php]: We should finally also introduce setters for these two members, shouldn't we?

File: src/main/php/scriptlet/HttpScriptlet.class.php [hunk 2]
```diff
@@ -426,24 +420,4 @@ public function service(HttpScriptletRequest $request, HttpScriptletResponse $re
...
-
-  /**
-   * This method is called to process any request and dispatches
-   * it to on of the do* -methods of the scriptlet. It will also
-   * call the <pre>doCreateSession()</pre> method if necessary.
-   *
-   * @return  scriptlet.HttpScriptletResponse the response object
-   * @throws  scriptlet.ScriptletException indicating fatal errors
-   */
-  public function process() {
```
Comment (kiesel) [inline src/main/php/scriptlet/HttpScriptlet.class.php]: Not sure here: this isn't marked as deprecated, is public and will be removed - no BC implications here? I know the preferred way of using a scriptlet is to use the `service()` method for a long time, but still...

**Comment by kiesel** [comment]
Like it; I am wondering if we need additional methods to enable deriving a specific layout in a project. Eg., it will be important to have the catch-all application be the last in the `mappedApplications` list, so it wouldn't be too greedy, but there's no way to handle that at the moment.

However, as this isn't prevented by this PR, I'd vote +1 :-)

File: src/main/php/xp/scriptlet/Runner.class.php (hunk 1 above)
Comment (thekid) [inline src/main/php/xp/scriptlet/Runner.class.php]: I'd really rather change that all the way to constructor parameters. For that, I'm currently working on getting rid of request and response subclassing in favor of wrapping.

In the future, we'd write:

```php
$request= new ScriptletRequest($method, $url, $headers, $env);
$wrapped= new XmlScriptletRequest($request);
```

File: src/main/php/scriptlet/HttpScriptlet.class.php (hunk 2 above)
Comment (thekid) [inline src/main/php/scriptlet/HttpScriptlet.class.php]: The only place it was used was `xp.scriptlet.Runner`, via `web[-main].php`; and the test suite.

**Comment by thekid** [comment]
>  it will be important to have the catch-all application be the last in the mappedApplications list

Yes, this is important: The regexes are matched in the order they're declared in the source. Just like in today's `web.ini`, by the way :grinning:

**Comment by thekid** [comment]
> but there's no way to handle that at the moment.

We could add a special `WebApplication defaultApplication($profile= null)` method and merge from the outside, however.

**Comment by thekid** [comment]
> but there's no way to handle that at the moment.

Another idea would be to add a constant `WebLayout ::DEFAULT` and sort the returned map by that before matching. 

```php
namespace com\example\web;

class Conventions extends Object implements WebLayout {
  private $package;

  public function __construct() {
    $conf= Properties::fromString(ClassLoader::getDefault()->getResource('project.ini'));
    $this->package= $conf->readString('project', 'package');
  }

  public function mappedApplications($profile= null) {
    $apps= [];
    $apps['/api']= (new WebApplication('api'))
      ->withConfig('{WEBROOT}/src/main/resources/{PROFILE}')
      ->withScriptlet('webservices.rest.srv.RestScriptlet')
      ->withArguments([$this->package.'.rest', '/api'])
     ;
    $apps[self::DEFAULT]= (new WebApplication('web'))
      ->withConfig('{WEBROOT}/src/main/resources/{PROFILE}')
      ->withScriptlet('com.example.web.commons.BaseXmlScriptlet')
      ->withArguments([$this->package.'.scriptlet', '{WEBROOT}/src/main/xsl'])
     ;
    return $apps;
  }

  public function staticResources($profile= null) {
    return ['^/static' => '{DOCUMENT_ROOT}'];
  }
}

**Pull request merged**